proposals/gateway-plusplus-phase1.md

```diff
@@ -0,0 +1,122 @@
+
+We should lean into this rather than push against it. In the short term, we can’t expect every user and developer to have a full implementation of the protocol. Strengthening the Gateway protocol to meet user needs is the path of least resistance and can enable numerous thin clients to be built for IPFS that can run in virtually any environment.
+
+This proposal establishes "Gateway++" as a meta-project for meeting these user needs with extensions to our gateway protocols and infra.
+
+## Problem Statement
+
+For reading data, the IPFS Gateway is already serving these users quite well. Not only does it allow them to read data from the IPFS network without running a full node, they are able to integrate with existing HTTP caching infrastructure to improve performance.
+
+However, the user needs for writing data to the network are not being met well by our current solutions. We have two interfaces that support writes over HTTP and neither was designed as a full solution to these user needs.
+
+- Pinning Service API.
+    - 😡 Requires you to already be running a full IPFS node as a client.
+    - 😡 Is not transactional. You have to poll in order to tell when the data is actually available.
+    - 😃 Is intended to be multi-tenant and authenticated using a flexible bearer token.
+- HTTP RPC API
+    - 😡 Is single-tenant. No authentication system.
+        - The RPC API is designed for a single user to manage a remote node, it was not designed and built to be multi-tenant and would look very different if it were.
+    - 😡 Is not always transactional. Many write calls return before they are available.
+
+These interfaces were built with particular user profiles in mind. The users we’re seeing in NFTs (including our own service as a user of IPFS) are not well served by these interfaces and we need a new “project” definition (Gateway++) that is centered around these user needs.
+
+Developers are building applications that have to run in a lot of constrained environments that can’t support a full node. Strengthening the Gateway to serve these needs over HTTP leverages endpoints they are already using and provides the features they need in a stateless protocol that has ubiquitous support.
+
+We are never going to hear “Sorry, I can’t use HTTP in this environment” from a developer.
+
+## Phase 1
+
+We should expect this to be a long term project with additional proposals in due time.
+
+For now, we have a list of high priority tasks that need to be fixed.
+
+In [nft.storage](http://nft.storage) we have the following high priority needs:
+
+- Add the Pinning API to ipfs-cluster.
+- Add transactional CAR file uploads to the Pinning API.
```

Comment:
My read: it should basically _be_ `/api/v0/dag/import`, but nicely integrated into the Pinning API. Send it a blob, receive back a success/fail message, and transactional in the sense that it's all imported or not. If the CAR has a problem part way through, then bail on all of it. Details on how to do this blob of binary should be resolved this week hopefully with the binary API discussion, is `multipart/form-data` appropriate here? If we're doing this fresh for the Pinning API then we have an opportunity to try out an alt approach that we might choose for a v0 binary solution.

Reply:
Cluster is adding support for CAR files on the /add endpoint (which otherwise mimics api/v0/add).

Some choices cluster (or I) made:

* It's a POST multipart - even if cluster just accepts a single CAR part with a single root, multipart is how we usually upload things in the web and it is flexible enough to do other things (like normal adding).
* CAR must have a single root. The cluster API is constrained by being able to Pin one thing, so CARs must have a single root, or otherwise multiple roots would have to be wrapped in a single CID. I see the pinning API also does not have a "multple pin" endpoint so this may be a reasonable limitation also in the pinning API.
* I did not add a new endpoint because there is significant overlap between adding CARs and adding files normally: replication factors, pin options, stream channels, pin sharding etc. If a Pinning API add endpoint is added for CARs, think it might be expanded in the future to do normal unixfs-adding, or raw block-adding.
* Cluster added a `format=<car/unixfs>` query option to the `/add` endpoint control how things are supposed to be added (choosing a DAG Formatter, which given an input produces ipld.Nodes as output).

Reply:
IIUC the pinning API was designed to handle the operation of pinning content which is a separate operation from pushing content to a particular endpoint. @lidel may be able to fill in the blanks here. I'm having trouble finding the slides at the moment, but this talk from Juan (and the slides in the background) https://youtu.be/Pcv8Bt4HMVU?t=912 setting the background for the pinning API discussion differentiates between the different types of operations that might need to be provided.

Using CAR files as a mechanism for pushing data is wasteful in that it ignores the existence of duplicate data at the endpoint. For example, adding a 10kB file to a 100MB directory now requires uploading 100MB of data. Making CAR file uploads "first class citizens" and the recommended way people interact with our stack is IMO a mistake.

Reply:
If there's actually high demand for HTTP-only environments having a standardized ingestion format that's not `/api/v0/import` then providing something here seems reasonable.

However, IMO we should have tooling in place that points people down a more correct path (i.e. a libp2p node that spins up a single WSS connection to the endpoint from the pinning API and sends the data over Bitswap/GraphSync). 

Additionally, it might be nice if we could allow people to be more efficient by being able to ask the pinning service "which blocks in this CAR file manifest do you already have?" and then only uploading a CAR with the delta of missing blocks. Since this is an optimization it can be done later if it's a pain.

Reply:
> Using CAR files as a mechanism for pushing data is wasteful in that it ignores the existence of duplicate data at the endpoint. 

This is true, but in practice people are most often not writing data that already exists on the node. In NFT’s, more than 99% of the data written will be unique. Having a fast transactional API for this would meet the primary user needs. Duplicate data on write is an edge case in the users we are focused on right now and should not drive the design of the interfaces.

Reply:
@mikeal I feel still confused, and I think it is  because there seem to be conflicting lines of reasoning here:
- on one hand, you argue that people don't want to run IPFS node for interacting with pinning service, and due to this pinning service should accept data upload
- on the other hand, the proposal is to... create CID for NFT, export it, then upload CAR archive, which by definition requires 80% of IPFS node to be present if one wants to create one

I mean.. I must be missing something here?

> Having a fast transactional API for this would meet the primary user needs.

If you use go-ipfs 0.8.0 node _today_, then you have transactional remote pinning over bitswap: 
- `pin remote service add` and `pin remote add` take care of interacting with the API
- go-ipfs takes care of [provider hints](https://ipfs.github.io/pinning-services-api-spec/#section/Provider-hints) and there is no need for waiting for DHT in most cases, bitswap session is started immediately
- `ipfs pin remote add` has  `--background=false`  mode which will wait until pinning is successful or fails – no need for writing polling code, go-ipfs takes care of it already

js-ipfs support for this is tracked in https://github.com/ipfs/js-ipfs/pull/3588. 

Sure, we want to make it possible to use Gateways for CAR and file imports, that will most likely happen, but...

...for this specific use case (NFT) the person minting the thing already has a full node somewhere, and that node has the data, because CID had to be created somehow.  

All you need to do is to leverage [provider hints](https://ipfs.github.io/pinning-services-api-spec/#section/Provider-hints)  built into the pinning service api spec, and let IPFS do its thing.

Reply:
> on the other hand, the proposal is to... create CID for NFT, export it, then upload CAR archive, which by definition requires 80% of IPFS node to be present if one wants to create one

This may have been true in the past but not anymore. We’ve made some big gains in these libraries, the codecs and unixfs encoding can be done with very little libraries now. `dag-pb` and `dag-cbor` are almost dependency free (we don’t actually need all of protobuf or CBOR since dag- is a strict subset).

The data encoding is quite small nowadays. What remains large are the storage layer interfaces, network interfaces, etc. Those are still quite large and difficult to untangle, but unnecessary if you’re just writing DAGs to a Gateway.

> transactional

Is there a `—background=false` equivalent in the HTTP pinning API?

Also, I’ve observed transactional integrity issues even in the block storage interface in js-ipfs https://github.com/mikeal/IPSQL/blob/cc33314410e15a271cb34e76c38edc4045128199/src/ipfs.js#L31-L49 but go-ipfs may be better about this.

Reply:
My first thought here was that the API would actually do the unixfs chunking for a single file for you, and return the CID, which would answer the dichotomy lidel proposed very directly. I want to `curl` a file to a gateway. But that needn't be necessarily coupled in here, since clearly we need the more general ability for DAGs.

I concur that just because our monolithic software todays means that in practise you need IPFS to chunk and CAR a file, we should not bake that in to APIs. In the future I see very thin clients that could chunk/CAR and upload, including web clients.

We're optimizing here for accessibility, ease of use and integration, and to that end I strongly support adding capabilities with HTTP. The possible gains in transmission of redundant blocks is besides the point. Operators for whom that is really important can choose the complicated path, everyone else can choose the easy path.

Reply:
Let me add that we should _also_ work towards software like @aschmahmann describes that would support a graphsync upload, but that would be more friction for the immediate needs.

Reply:
@hsanjuan can you confirm that this item is _done_ as of https://github.com/ipfs/ipfs-cluster/pull/1343?
So the only thing left in this proposal is the pinning API to cluster? (plus the doc+deploy items below)
Or is there more to do with CAR uploads?

Reply:
at the very least i’d expect the CAR upload feature to need to be updated to accept and validate a token in the same way the Pinning API does after the Pinning API lands in cluster.

Reply:
> @hsanjuan can you confirm that this item is done as of ipfs/ipfs-cluster#1343?

The item as written is not done. Cluster added CAR-file import to its own REST API which is different than the official Pinning API (which it does not have). When the Pinning API knows how it wants to support CAR file import, it should be easy to re-use the importer that cluster includes now, along with the rest of the Pinning API and the token-based authentication.